src/differentiable_functions.jl

```diff
@@ -1,91 +1,159 @@
 abstract type AbstractDifferentiableMultivariateFunction end
 
-struct DifferentiableMultivariateFunction <: AbstractDifferentiableMultivariateFunction
-    f!::Function
-    g!::Function
-    fg!::Function
+struct DifferentiableMultivariateFunction{F1,F2,F3} <: AbstractDifferentiableMultivariateFunction
+    f!::F1
+    g!::F2
+    fg!::F3
+
+    DifferentiableMultivariateFunction{F1,F2,F3}(f!, g!, fg!) where {F1,F2,F3} =
+        new{F1,F2,F3}(f!, g!, fg!)
 end
 
 alloc_jacobian(df::DifferentiableMultivariateFunction, T::Type, n::Integer) = Array{T}(n, n)
 
-function DifferentiableMultivariateFunction(f!::Function, g!::Function)
-    function fg!(x::Vector, fx::Vector, gx::Array)
+DifferentiableMultivariateFunction(f!, g!, fg!) =
+    DifferentiableMultivariateFunction{typeof(f!),typeof(g!),typeof(fg!)}(f!, g!, fg!)
+
+function DifferentiableMultivariateFunction(f!, g!, initial_x::AbstractArray)
+    return DifferentiableMultivariateFunction(reshape_f(f!, initial_x),
+                                              reshape_g(g!, initial_x))
+end
+
+function DifferentiableMultivariateFunction(f!, g!, fg!, initial_x::AbstractArray)
+    return DifferentiableMultivariateFunction(reshape_f(f!, initial_x),
+                                              reshape_g(g!, initial_x),
+                                              reshape_fg(fg!, initial_x))
+end
+
+function DifferentiableMultivariateFunction(f!, g!)
+    function fg!(x::AbstractVector, fx::AbstractVector, gx::AbstractMatrix)
         f!(x, fx)
         g!(x, gx)
     end
     return DifferentiableMultivariateFunction(f!, g!, fg!)
 end
 
-function DifferentiableMultivariateFunction(f!::Function; dtype::Symbol=:central)
-    function fg!(x::Vector, fx::Vector, gx::Array)
+function DifferentiableMultivariateFunction(f!; dtype::Symbol=:central)
+    function fg!(x::AbstractVector, fx::AbstractVector, gx::AbstractMatrix)
```

Comment:
again, I'm wondering if this is really needed? I know the annotations were there, but...

Reply:
No, it's not I guess though I haven't tried to remove the annotations. Most/all annotations are rather reminders and assertions to both show and ensure which type of function is created or expected. Do you want to get rid of annotations where possible?